README
Graph::Flowchart
================

This module helps you to assemble flowcharts for code.
It will generate Graph::Easy objects, which can then be output in various
formats like HTML, ASCII, Unicode boxart drawings, graphviz or SVG.

As an example look for Devel::Graph, which uses this package to turn this:

	my $a = shift;
	my $b = 1;
	if ($a < 9)
	  {
	  $b = $a + 1;
	  }
	return $b + 2;

into a Graph::Easy object, which lets you then generate output like
the following:

	##################
	#     start      #
	##################
	  |
	  |
	  v
	+----------------+
	| my $a = shift; |
	| my $b = 1;     |
	+----------------+
	  |
	  |
	  v
	+----------------+
	|  if ($a < 9)   |--+
	+----------------+  |
	  |                 |
	  | true            |
	  v                 |
	+----------------+  |
	|  $b = $a + 1;  |  |
	+----------------+  |
	  |                 |
	  |                 | false
	  v                 |
	+----------------+  |
	| return $b + 2; | <+
	+----------------+
	  |
	  |
	  v
	##################
	#      end       #
	##################


INSTALLATION
============

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Copyright (C) 2004-2005 by Tels C<http://bloodgate.com>

See the LICENSE file for more information.



CHANGES
Revision history for Perl extension Graph::Flowchart:

0.10 2007-02-02 Tels 36 tests
  * require Perl v5.8.1
  * require Graph::Easy v0.52
  * bundle and use Module::Install v0.64
  * add start_node(), end_node()
  * finish(): remove start/end node if they are isolated and only connected
    to each other, unless they are the only nodes

0.09 2006-01-29 Tels 36 tests
  * fix Makefile.PL to work with newer YAML
  * require Graph::Easy v0.40

0.08 2006-01-08 Tels 36 tests
  * require Graph::Easy v0.39 (to correctly be able to merge blocks)
  * fix an undef warning in line 321 
  * fix the conversion of "\n" to "\\n" to actually work
  * add N_WHILE, N_UNTIL types
  * add: insert_block(), insert_new_block(), insert_new_joint()
  * document that these are not working when used recursive:
    + add_while
    + add_until
    + add_foreach
    + add_for
    + add_if_then
    + add_if_then_else

0.07 2006-01-01 Tels 36 tests
  * require Graph::Easy v0.38
  * set edge flow to better handle the flow after loops
  * use constant;
  * add: add_until(), add_foreach()
  * add_for() returns (current, body, continue) blocks

0.06 2005-12-27 Tels 36 tests
  * require Graph::Easy v0.37
  * add:
    + add_goto(), add_break(), add_return(), add_last(), add_new_block()
    + make_current(), current(), add_new_join(),
    + N_GOTO N_RETURN N_BREAK N_NEXT N_LAST N_SUB N_USE
  * connect() can set the edge class, too
  * set edge classes for: true, false, goto, return, break, last, next
  * remove dead code relating to setting graph attributes
  * new_block(): handle label names (for goto)
  * Node.pm: turn '\n' into '\n' for correct node labels
  * fix POD of Node.pm
  * use strict, not warnings in .pm files
  * add_if_then() and add_if_then_else() return all blocks in list context

0.05 2005-12-19 Tels 23 tests
  * require Graph::Easy v0.36
  * set a subclass for each node
  * set attributes on a per-class basis
  * fix pod.t and pod_cov.t when Pod::Test/Pod::Coverage are not installed
  * fix the POD in both .pm files
  * fix README and include example
  * add_if_then_else() can handle undef else block
  * first_block(), last_block() and current_block() can now also set the block

0.04 2005-12-07 Tels 19 tests
  * taken code over from Devel::Graph
  * require Graph::Easy v0.34
  * fix to work with Graph::Easy 0.34
  * make pod tests also cover ::Node
  * fix graph detailing usage of merge()

0.03 2005-11-13 Tels 17 tests
  * require Graph::Easy v0.33
  * require PPI v1.04
  * require Perl v5.8.0
  * update POD with examples and note that decomposing is not implemented yet
  * add: add_while()

0.02 2005-11-06 Tels 18 tests
  * require Graph::Easy 0.32

0.01 2005-11-06 Tels 18 tests
  * first version



examples/test.pl
#!/usr/bin/perl -w

use lib '../lib';
use Graph::Flowchart;

my $format = shift || 'as_boxart';

my $g = Graph::Flowchart->new();

$g->add_block ('$a = "9";');
$g->add_block ('my $b = 1;');
$g->add_if_then ( 'if ($a == 9)', '$b == 9;' );
$g->add_if_then_else ( 'if ($b == 9)', '$b == $a + 1;', '$c == 1' );

$g->add_for ( 'my $i = 0;', 'for: $i < 10;', '$i++', '$a++;');

$g->add_for ( 'my $i = 0;', 'for: $i < 10;', '$i++', undef );

$g->add_while ( 'while ($b < 19)', '$b++;' );		# no continue block
$g->add_while ( 'while ($b < 22)', undef, '$b++;' );	# no body block
$g->add_while ( 'while ($b < 24)', '$a++;', '$b++;' );	# both body&continue

$g->finish();

my $gr = $g->as_graph();

#$gr->set_attribute('flow','right');

print STDERR "Resulting graph has ", 
	scalar $gr->nodes(), " nodes and ", 
	scalar $gr->edges()," edges:\n\n";

binmode STDOUT, ':utf8' or die ("binmode STDOUT, ':utf8' failed: $!");
print $gr->$format();



lib/Graph/Flowchart/Node.pm
#############################################################################
# A node in the graph (includes a type field)
#
# (c) by Tels 2004-2005.
#############################################################################

package Graph::Flowchart::Node;

@ISA = qw/Graph::Easy::Node Exporter/;
$VERSION = '0.06';

use Graph::Easy::Node;
use Exporter;

@EXPORT_OK = qw/
  N_START N_END N_BLOCK N_IF N_THEN N_ELSE N_JOINT N_END N_FOR N_BODY
  N_USE
  N_SUB
  N_CONTINUE N_GOTO N_BREAK N_RETURN N_NEXT N_LAST N_WHILE N_UNTIL
  /;

#############################################################################
#############################################################################

use strict;

use constant {
  N_START	=> 1,
  N_END		=> 2,
  N_BLOCK	=> 3,
  N_IF		=> 4,
  N_THEN	=> 5,
  N_ELSE	=> 6,
  N_JOINT	=> 7,
  N_FOR		=> 8,
  N_BODY	=> 9,
  N_CONTINUE	=> 10,
  N_GOTO	=> 11,
  N_RETURN	=> 12,
  N_BREAK	=> 13,
  N_NEXT	=> 14,
  N_LAST	=> 15,
  N_SUB		=> 16,
  N_USE		=> 17,
  N_WHILE	=> 18,
  N_UNTIL	=> 19,
  };

my $subclass = {
  N_START()	=> 'start',
  N_END()	=> 'end',
  N_BLOCK()	=> 'block',
  N_BODY()	=> 'block',
  N_CONTINUE()	=> 'block',
  N_THEN()	=> 'block',
  N_IF()	=> 'if',
  N_ELSE()	=> 'else',
  N_JOINT()	=> 'joint',
  N_FOR()	=> 'for',
  N_GOTO()	=> 'goto',
  N_RETURN()	=> 'return',
  N_NEXT()	=> 'next',
  N_BREAK()	=> 'break',
  N_LAST()	=> 'last',
  N_SUB()	=> 'sub',
  N_USE()	=> 'use',
  N_WHILE()	=> 'while',
  N_UNTIL()	=> 'until',
  };

#############################################################################
#############################################################################

sub new
  {
  my ($class, $label, $type, $labelname, $group) = @_;

  my $self = bless {}, $class;

  $type = N_START() unless defined $type;

  $self->{_type} = $type;
  $self->{_label} = $labelname if defined $labelname;
  $self->{id} = Graph::Easy::Base::_new_id();

  # convert newlines into '\n'
  $label =~ s/\n/\\n/g;

  $self->_init( { label => $label, name => $self->{id} } );

  $self->sub_class($subclass->{$type} || 'unknown');

  if (ref $group)
    {
    $group->add_node($self);
    }

  $self;
  }

sub _set_type
  {
  my $self = shift;

  $self->{_type} = shift;

  $self;
  }

1;
__END__

=head1 NAME

Graph::Flowchart::Node - A node in a Graph::Flowchart, representing a block/expression

=head1 SYNOPSIS

	use Devel::Graph;

	my $graph = Devel::Graph->graph( '$a = 9 if $b == 1' );

	print $graph->as_ascii();

=head1 DESCRIPTION

This module is used by Devel::Graph internally, there should be no need to
use it directly.

=head2 EXPORT

Exports nothing on default but can export on request the following:

  N_START N_END N_BLOCK N_IF N_THEN N_ELSE N_JOINT N_END N_FOR N_BODY
  N_SUB
  N_CONTINUE N_GOTO N_RETURN N_BREAK N_NEXT N_LAST
  N_USE

=head1 METHODS

=head2 new()

	my $node = Graph::Flowchart::Node->new( $text, $type, $label);
	my $node = Graph::Flowchart::Node->new( $text, $type);

Create a new node of the given C<$type> with the C<$text> as label. The
optional C<$label> is the label for C<goto>.

For instance:

	LABEL: $a = 1;
	goto LABEL;

would be turned into:

	my $n1 = Graph::Flowchart::Node->new( "$a = 1;\n", N_BLOCK, 'LABEL'); 
	my $n2 = Graph::Flowchart::Node->new( "goto LABEL;\n", N_GOTO); 

=head1 SEE ALSO

L<Devel::Graph>, <Graph::Easy>.

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms of the GPL version 2.
See the LICENSE file for information.

X<gpl>

=head1 AUTHOR

Copyright (C) 2004-2006 by Tels L<http://bloodgate.com>

X<tels>

=cut
